### DISCUSSION TITLE
docs: asyncio start_server does not behave as described

### DISCUSSION THREAD

**Discussion by byzantic** [micropython/micropython#13592] [Libraries & Drivers]
Micropython version: 1.22.1 
Board: W5500-EVB-Pico

Wiznet has a board that adds ethernet capability to a Raspberry Pi Pico - the W5500-EVB-Pico. I'd like to use this to implement network connected instruments in my home lab.

There are plenty of examples on the documentation site, but the TCP examples all seem to use low-level socket programming. The latest Micropython implementations all have asyncio available so why not use the facilities available there?

For example [Bytepawn's simple message queue server](https://bytepawn.com/writing-a-simple-python-async-message-queue-server.html).  Here is a snippet that shows how `start_server` is used:

```python
async def run_server(host, port):
    server = await asyncio.start_server(handle_client, host, port)
    print(f'Listening on {host}:{port}...')
    async with server:
        await server.serve_forever()

asyncio.run(run_server(host='localhost', port=int(sys.argv[1])))
```
Nope. Doesn't work. The problem seems to be that even though the Micropython documentation says that the asyncio.start_server function works in the same way as the standard library - it doesn't. Specifically, the documentation states that `start_server` returns an instance of the `Server` class. It doesn't - it returns a `<generator>`.

After some digging on the net, I came up with the code below (mainly through the infinite monkey procedure).

This code uses start_server but then runs the resulting generator using an explicitly constructed `asyncio` `event_loop`. There is an advantage to this, though, in that this also allows us to start multiple tasks, or multiple servers at the same time.

The create_handler function builds a parametrised handler that is passed to start_server. It can be parametrised with a function that transforms from a string to a string. The default is just the identity function.

Run the program and use telnet to connect to to the board on either port 7777 - which acts as an echo server, or port 7778 which uppercases the reply.

```python
import  sys, asyncio

#--------------------------------------------------------------------------
# W5500-EVB-Pico board and network configuration
#--------------------------------------------------------------------------
# on micropython systems, this can be placed in boot.py

from machine import Pin,SPI
import network
import time

#W5x00 interface configuration
def config():
    spi=SPI(0,2_000_000, mosi=Pin(19),miso=Pin(16),sck=Pin(18))
    nic = network.WIZNET5K(spi,Pin(17),Pin(20)) #spi,cs,reset pin
    nic.active(True)
    
    #Use if static IP address is required
    #nic.ifconfig(('192.168.1.20','255.255.255.0','192.168.1.1','8.8.8.8'))
    
    #DHCP
    nic.ifconfig('dhcp')
    print('IP address :', nic.ifconfig())
    
    while not nic.isconnected():
        time.sleep(1)
        print(nic.regs())

#--------------------------------------------------------------------------
# END network configuration
#--------------------------------------------------------------------------
        
async def handle_client(reader, writer):
    print('New client connected...')
    line = str()
    while line.strip() != 'quit':
        line = (await reader.readline()).decode('utf8')
        if line.strip() == '': continue
        print(f'Received: {line.strip()}')
        writer.write(line.encode('utf8'))
    writer.close()
    print('Client disconnected...')

def create_handler(transformer = lambda x : x):
    async def handle_client(reader, writer):
        print('New client connected...')
        line = str()
        while line.strip() != 'quit':
            line = (await reader.readline()).decode('utf8')
            if line.strip() == '': continue
            print(f'Received: {line.strip()}')
            writer.write(transformer(line).encode('utf8'))
        writer.close()
        print('Client disconnected...')
    return handle_client

async def Main():
    config()
    loop = asyncio.get_event_loop()
    loop.create_task(asyncio.start_server(create_handler(),           '0.0.0.0',7777))
    loop.create_task(asyncio.start_server(create_handler(str.upper),  '0.0.0.0',7778))
    loop.run_forever()

try:
    asyncio.run(Main())
finally:
    asyncio.new_event_loop()
```

**Comment by glenn20** [comment]
Micropython does not support the `Server.serve_forever()` method, which I would have expected to be apparent from an error message when running the code.

> The problem seems to be that even though the Micropython documentation says that the asyncio.start_server function works in the same way as the standard library - it doesn't. Specifically, the documentation states that start_server returns an instance of the Server class. It doesn't - it returns a <generator>.

In fact **all** awaitable functions technically return a generator (that is how asyncio is implemented in python). The docs refer to what is returned from `await`-ing the generator returned by an awaitable.

I **believe** the original code will work if you replace `await server.serve_forever()` with `while True: await asyncio.sleep(300)` or some other form of async wait forever.
